File: snippets/cross-host-communication/python/larry_telemetry.py
"""Larry telemetry publisher."""

# snippet:start
import iceoryx2 as iox2
from telemetry_data import BatteryState, Position

cycle_time = iox2.Duration.from_secs(1)

iox2.set_log_level_from_env_or(iox2.LogLevel.Info)
node = iox2.NodeBuilder.new().create(iox2.ServiceType.Ipc)

battery_publisher = (
    node.service_builder(iox2.ServiceName.new("larry/battery"))
    .publish_subscribe(BatteryState)
    .open_or_create()
    .publisher_builder()
    .create()
)
battery_notifier = (
    node.service_builder(iox2.ServiceName.new("larry/battery"))
    .event()
    .open_or_create()
    .notifier_builder()
    .create()
)

position_publisher = (
    node.service_builder(iox2.ServiceName.new("larry/position"))
    .publish_subscribe(Position)
    .open_or_create()
    .publisher_builder()
    .create()
)
position_notifier = (
    node.service_builder(iox2.ServiceName.new("larry/position"))
    .event()
    .open_or_create()
    .notifier_builder()
    .create()
)

try:
    while True:
        node.wait(cycle_time)

        battery = battery_publisher.loan_uninit()
        battery = battery.write_payload(BatteryState(charge_percent=87.0))
        battery.send()
        battery_notifier.notify()

        position = position_publisher.loan_uninit()
        position = position.write_payload(Position(x=1.0, y=2.0))
        position.send()
        position_notifier.notify()

except iox2.NodeWaitFailure:
    print("exit")
# snippet:end

File: snippets/cross-host-communication/rust/src/bin/larry_telemetry.rs
fn main() -> Result<(), Box<dyn core::error::Error>> {
    // snippet:start
    use core::time::Duration;
    use cross_host_communication::telemetry_data::{BatteryState, Position};
    use iceoryx2::prelude::*;

    const CYCLE_TIME: Duration = Duration::from_secs(1);

    let node = NodeBuilder::new()
        .name(&"Larry".try_into()?)
        .create::<ipc::Service>()?;

    let battery_publisher = node
        .service_builder(&"larry/battery".try_into()?)
        .publish_subscribe::<BatteryState>()
        .open_or_create()?
        .publisher_builder()
        .create()?;
    let battery_notifier = node
        .service_builder(&"larry/battery".try_into()?)
        .event()
        .open_or_create()?
        .notifier_builder()
        .create()?;

    let position_publisher = node
        .service_builder(&"larry/position".try_into()?)
        .publish_subscribe::<Position>()
        .open_or_create()?
        .publisher_builder()
        .create()?;
    let position_notifier = node
        .service_builder(&"larry/position".try_into()?)
        .event()
        .open_or_create()?
        .notifier_builder()
        .create()?;

    while node.wait(CYCLE_TIME).is_ok() {
        let battery = battery_publisher.loan_uninit()?;
        let battery = battery.write_payload(BatteryState {
            charge_percent: 87.0,
        });
        battery.send()?;
        battery_notifier.notify()?;

        let position = position_publisher.loan_uninit()?;
        let position = position.write_payload(Position { x: 1.0, y: 2.0 });
        position.send()?;
        position_notifier.notify()?;
    }
    // snippet:end

    Ok(())
}
